Dockerfile

```diff
@@ -1,59 +1,81 @@
-    && mkdir -p /etc/coraza-spoa/rules \
-    && tar xzf /tmp/coreruleset.tar.gz --strip-components=1 -C /tmp/coraza-coreruleset \
-    && mv /tmp/coraza-coreruleset/crs-setup.conf.example /etc/coraza-spoa/crs-setup.conf \
-    && mv /tmp/coraza-coreruleset/rules /etc/coraza-spoa \
-    && mv /tmp/coraza-coreruleset/plugins /etc/coraza-spoa \
-    && rm -rf /tmp/*
-
-# Cache Go dependencies
-COPY go.mod go.sum ./
+WORKDIR /build
+COPY . /build
+
+# Download dependencies for all platforms once
 RUN go mod download
 
-# Copy project files into build image
-COPY . .
+ARG TARGETOS
+ARG TARGETARCH
+
+RUN apk add --no-cache make ca-certificates \
+    && update-ca-certificates
+
+RUN --mount=type=cache,target=/root/.cache/go-build \
+    --mount=type=cache,target=/go/pkg \
+    OS=${TARGETOS} ARCH=${TARGETARCH} make
+
+# ---
+FROM alpine:3.17 AS main
+
+ARG TARGETARCH
+
+LABEL org.opencontainers.image.authors="OWASP Coraza Maintainers" \
+      org.opencontainers.image.description="OWASP Coraza WAF (Haproxy SPOA)" \
+      org.opencontainers.image.documentation="https://coraza.io/connectors/coraza-spoa/" \
+      org.opencontainers.image.licenses="Apache-2.0" \
+      org.opencontainers.image.source="https://github.com/corazawaf/coraza-spoa" \
+      org.opencontainers.image.title="coraza-spoa"
+
+RUN apk add --no-cache tini socat ca-certificates \
+    && update-ca-certificates
+
+# Add unprivileged user & group for the coraza-spoa
+RUN addgroup --system coraza-spoa \
+    && adduser --system --ingroup coraza-spoa --no-create-home --home /nonexistent --disabled-password coraza-spoa
 
-# Build coraza-spoa binary
-RUN make
+RUN mkdir -p /etc/coraza-spoa /var/log/coraza-spoa \
+    && chown coraza-spoa:coraza-spoa /var/log/coraza-spoa
+
+COPY --from=builder /build/coraza-spoa_${TARGETARCH} /usr/bin/coraza-spoa
+COPY --from=builder /build/config.yaml.default /etc/coraza-spoa/config.yaml
+COPY --from=builder /build/docker/coraza-spoa/coraza.conf /etc/coraza-spoa/coraza.conf
+COPY --from=builder /build/docker/coraza-spoa/docker-entrypoint.sh /docker-entrypoint.sh
+
+EXPOSE 9000
+USER coraza-spoa
+
+HEALTHCHECK --interval=10s --timeout=2s --retries=2 CMD "/usr/bin/socat /dev/null TCP:0.0.0.0:9000"
+
+ENTRYPOINT ["tini", "--", "/docker-entrypoint.sh"]
+
+CMD ["/usr/bin/coraza-spoa", "--config", "/etc/coraza-spoa/config.yaml"]
+
+# ---
+FROM main AS coreruleset
+
+ARG CORERULESET_VERSION=v4.0.0-rc1
+ARG CORERULESET_SHA256SUM=a8f0d1cac941bf2158988b92a91519f093a8bce64a260e46fa352d608c7de3e6
 
-RUN \
-    # Move coraza-spoa config file to config dir
-    mv /app/docker/coraza-spoa/coraza.conf /etc/coraza-spoa/coraza.conf \
-    # Rename coraza-spoa default config file
-    && mv /app/config.yaml.default /app/config.yaml \
-    # Rename coraza-spoa binary
-    && mv /app/coraza-spoa_amd64 /app/coraza-spoa \
-    # Make coraza-spoa binary executable
-    && chmod +x /app/coraza-spoa
+# Switch to root for crs installation
+USER root
 
+# Download the core rule set
```

Comment:
While this is outside the scope of the PR in the future we might want to use coraza-coreruleset.